oauth_dropins/meetup.py

```diff
@@ -0,0 +1,162 @@
+        """
+        return self.key.string_id()
+
+
+def access_token(self):
+    """Returns the OAuth access token string.
+    """
+    return self.access_token_str
+
+
+def urlopen(self, url, **kwargs):
+    """Wraps urllib2.urlopen() and adds OAuth credentials to the request.
+    """
+    headers = {'Authorization': 'Bearer %s' % self.access_token_str}
+    try:
+        return util.urlopen(urllib2.Request(url, headers=headers), **kwargs)
+    except BaseException as e:
+        util.interpret_http_exception(e)
+        raise
+
+
+class MeetupCsrf(ndb.Model):
+    """Stores a CSRF token for the Meetup.com OAuth2 flow."""
+    token = ndb.StringProperty(required=False)
+    state = ndb.TextProperty(required=False)
+
+
+class StartHandler(handlers.StartHandler):
+    """Starts Meetup.com auth. Requests an auth code and expects a redirect back.
+    """
+    def redirect_url(self, state=None):
+        assert (MEETUP_CLIENT_ID and MEETUP_CLIENT_SECRET), (
+                "Please fill in the meetup_client_id and meetup_client_secret files in "
+                "your app's root directory.")
+
+        csrf_key = MeetupCsrf(state=state).put()
+        return GET_AUTH_CODE_URL % {
+                'client_id': MEETUP_CLIENT_ID,
+                'redirect_uri': urllib.quote_plus(self.to_url(state=state)),
+                'state': '%s|%s' % (state, csrf_key.id()),
+                }
+
+
+class CallbackHandler(handlers.CallbackHandler):
+    """The auth callback. Fetches an access token, stores it, and redirects home.
+    """
+    def get(self):
+      # handle errors
+      error = self.request.get('error')
+      if error:
+          if error == 'access_denied':
+              logging.info('User declined')
+              self.finish(None, state=self.request.get('state'))
+              return
+          else:
+              msg = 'Error: %s: %s' % (error, self.request.get('error_description'))
+              logging.info(msg)
+              raise exc.HTTPBadRequest(msg)
+
+      state = util.get_required_param(self, 'state')
+      # lookup the CSRF token
+      try:
+          csrf_id = int(urllib.parse.unquote_plus(state).split('|')[-1])
+      except (ValueError, TypeError):
+          raise exc.HTTPBadRequest('Invalid state value %r' % state)
+
+      csrf = MeetupCsrf.get_by_id(csrf_id)
+      if not csrf:
+          raise exc.HTTPBadRequest('No CSRF token for id %s' % csrf_id)
+
+      # extract auth code and request access token
+      auth_code = util.get_required_param(self, 'code')
+      data = {
+          'code': auth_code,
+          'client_id': MEETUP_CLIENT_ID,
+          'client_secret': MEETUP_CLIENT_SECRET,
+          # redirect_uri here must be the same in the oauth code request!
+          # (the value here doesn't actually matter since it's requested server side.)
+          'redirect_uri': self.request.path_url,
+          }
+      resp = util.requests_post(GET_ACCESS_TOKEN_URL,
```

Comment:
you'll also want to `resp.raise_for_status()` here before using `.text` to catch and raise non-2xx response codes. the base handler class uses `webutil.handlers.handle_exception()` to catch requests' exceptions and convert them to appropriate HTTP responses.